x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java

```diff
@@ -1137,6 +1142,51 @@ protected Expression rule(Expression e) {
         }
     }
 
+    public static final class BubbleUpNegations extends OptimizerExpressionRule {
```

Comment:
We left out the negations from the discussed operators. Sorry for this late discovery, but just realized that the `Integer.{MIN,MAX}_VALUE` differences would bite us (small chance, but still):

```
-- with the negations in place, this works
SELECT -1073741824 * 2 / -3

-- after optimizing the negations this blows up with integer overflow
SELECT 1073741824 * 2 / 3
```

Reply:
Correct, thanks for pointing it out.

I guess a more immediate example is simply `-int * 2`, the before/after difference being the working `NEG(int) * 2` vs the failing `NEG(int * 2)` for the above "-MIN_VALUE/2" value.